File: clients/intellij-community/src/main/java/dev/cleanroommc/workbench/intellij/community/MaterialDeliveryClient.java
package dev.cleanroommc.workbench.intellij.community;

import com.google.gson.JsonObject;
import java.util.List;
import static dev.cleanroommc.workbench.intellij.community.MaterialChecksClient.text;

/** Identity and paging guards for diagnostics retained before snapshot finalization. */
final class MaterialDeliveryClient {
    static final String VIEW = "workbench-material-diagnostic-view-v1";
    static final String STATUS = "workbench-material-delivery-status-v1";
    static final List<String> GROUPS = List.of("error-located", "error-unlocated", "warning-located", "warning-unlocated", "information-located", "information-unlocated");
    private MaterialDeliveryClient() { }
    static boolean isView(JsonObject value) { return VIEW.equals(text(value, "format", "")); }
    static boolean revision(String value) { return value.matches("check-diagnostics:sha256:[0-9a-f]{64}"); }
    static JsonObject options(JsonObject view) {
        var options = new JsonObject(); options.add("revision", view.get("diagnostic_id")); return options;
    }
    static void arguments(List<String> args, String action, JsonObject options) {
        if (options != null && options.has("revision")) {
            if (!revision(text(options, "revision", ""))) throw new IllegalArgumentException("Select an exact diagnostic revision.");
            args.addAll(List.of("--revision", options.get("revision").getAsString()));
        }
        if (action.equals("diagnostics")) {
            long offset = options != null && options.has("offset") ? options.get("offset").getAsLong() : 0;
            if (offset < 0) throw new IllegalArgumentException("Select an exact diagnostic offset.");
            args.addAll(List.of("--offset", Long.toString(offset)));
            if (options != null && options.has("group")) {
                if (!GROUPS.contains(text(options, "group", ""))) throw new IllegalArgumentException("Select an exact diagnostic group.");
                args.addAll(List.of("--group", options.get("group").getAsString()));
            }
        }
        if (action.equals("diagnostic")) {
            if (options == null || !revision(text(options, "revision", "")) || !text(options, "finding", "").matches("diagnostic-(0|[1-9][0-9]*)"))
                throw new IllegalArgumentException("Select an exact retained diagnostic.");
            args.addAll(List.of("--diagnostic", options.get("finding").getAsString()));
        }
    }
    static void validate(JsonObject result, String action, JsonObject options) {
        if (STATUS.equals(text(result, "format", ""))) {
            if (!result.has("diagnostic_id") || !result.get("diagnostic_id").isJsonNull() && !revision(text(result, "diagnostic_id", ""))
                    || !List.of("ready", "preparing", "interrupted", "not-started").contains(text(result, "detail_state", "")))
                throw new IllegalArgumentException("Invalid diagnostic delivery status.");
        }
        if (isView(result)) {
            long offset = options != null && options.has("offset") ? options.get("offset").getAsLong() : 0;
            String group = text(result, "group", ""), selected = text(options, "group", "");
            if (!group.equals(selected)) throw new IllegalArgumentException("Diagnostic response belongs to another group.");
            long global = result.get("findings_count").getAsLong();
            long total = result.has("group_count") ? result.get("group_count").getAsLong() : global;
            if (!group.isEmpty() && (!GROUPS.contains(group) || !result.has("finding_counts"))) throw new IllegalArgumentException("Diagnostic group counts are unavailable.");
            if (result.has("finding_counts")) {
                var counts = result.getAsJsonObject("finding_counts");
                long sum = 0;
                if (counts.size() != GROUPS.size()) throw new IllegalArgumentException("Diagnostic groups differ.");
                for (String key : GROUPS) {
                    if (!counts.has(key) || !counts.get(key).toString().matches("0|[1-9][0-9]*")) throw new IllegalArgumentException("Invalid diagnostic group count.");
                    sum = Math.addExact(sum, counts.get(key).getAsLong());
                }
                if (sum != global || total != (group.isEmpty() ? global : counts.get(group).getAsLong())) throw new IllegalArgumentException("Diagnostic group counts differ from total findings.");
            }
            long count = result.getAsJsonArray("findings").size(), end = offset + count;
            if (!revision(text(result, "id", "")) || !result.get("id").equals(result.get("diagnostic_id"))
                    || !result.get("id").equals(result.get("view_id")) || result.get("offset").getAsLong() != offset || offset < 0 || total < end || total > global
                    || (result.get("next_offset").isJsonNull() ? end != total : count == 0 || end >= total || result.get("next_offset").getAsLong() != end))
                throw new IllegalArgumentException("Diagnostic page differs from its count or offset.");
        }
        if (options != null && options.has("revision") && List.of("diagnostics", "diagnostic").contains(action)
                && !options.get("revision").equals(result.get("diagnostic_id"))) throw new IllegalArgumentException("Diagnostic response belongs to another revision.");
        if (action.equals("diagnostic") && !options.get("finding").equals(result.getAsJsonObject("finding").get("id")))
            throw new IllegalArgumentException("Diagnostic response belongs to another finding.");
    }
}

File: clients/vscode/materialDeliveryClient.js
"use strict";
const VIEW = "workbench-material-diagnostic-view-v1";
const STATUS = "workbench-material-delivery-status-v1";
const revisionPattern = /^check-diagnostics:sha256:[0-9a-f]{64}$/;
const GROUPS = ["error-located", "error-unlocated", "warning-located", "warning-unlocated", "information-located", "information-unlocated"];
function argumentsFor(action, value) {
  const args = [];
  if (value?.revision !== undefined) {
    if (!revisionPattern.test(value.revision)) throw new Error("Select an exact diagnostic revision.");
    args.push("--revision", value.revision);
  }
  if (action === "diagnostics") {
    if (!Number.isSafeInteger(value?.offset ?? 0) || (value?.offset ?? 0) < 0) throw new Error("Select an exact diagnostic offset.");
    args.push("--offset", String(value?.offset ?? 0));
    if (value?.group !== undefined) {
      if (!GROUPS.includes(value.group)) throw new Error("Select an exact diagnostic group.");
      args.push("--group", value.group);
    }
  }
  if (action === "diagnostic" || action === "source") {
    if (!revisionPattern.test(value?.revision || "") || !/^diagnostic-(0|[1-9][0-9]*)$/.test(value?.finding || "")) throw new Error("Select an exact retained diagnostic.");
    args.push("--diagnostic", value.finding);
  }
  return args;
}
function validate(result, action, options) {
  if (result.format === STATUS) {
    if (result.diagnostic_id !== null && !revisionPattern.test(result.diagnostic_id || "")) throw new Error("Invalid diagnostic revision.");
    if (!["ready", "preparing", "interrupted", "not-started"].includes(result.detail_state)) throw new Error("Invalid delivery state.");
  }
  if (result.format === VIEW) {
    const total = result.group_count ?? result.findings_count;
    if ((result.group ?? null) !== (options?.group ?? null)) throw new Error("Diagnostic response belongs to another group.");
    if (result.group && (!GROUPS.includes(result.group) || !result.finding_counts)) throw new Error("Diagnostic group counts are unavailable.");
    if (result.finding_counts) {
      if (Object.keys(result.finding_counts).length !== GROUPS.length
          || GROUPS.some(group => !Number.isSafeInteger(result.finding_counts[group]) || result.finding_counts[group] < 0)
          || GROUPS.reduce((n, group) => n + result.finding_counts[group], 0) !== result.findings_count
          || total !== (result.group ? result.finding_counts[result.group] : result.findings_count)) throw new Error("Diagnostic group counts differ from total findings.");
    }
    if (!revisionPattern.test(result.id || "") || result.diagnostic_id !== result.id || result.view_id !== result.id
        || !Array.isArray(result.findings) || !Number.isSafeInteger(result.findings_count) || result.findings_count < 0
        || !Number.isSafeInteger(total) || total < 0 || total > result.findings_count
        || result.offset !== (options?.offset ?? 0) || result.offset + result.findings.length > total
        || result.next_offset !== (result.offset + result.findings.length === total ? null : result.offset + result.findings.length)
        || (result.next_offset !== null && result.findings.length === 0)) throw new Error("Diagnostic page differs from its count or offset.");
  }
  if (options?.revision && ["diagnostics", "diagnostic"].includes(action) && result.diagnostic_id !== options.revision) throw new Error("Diagnostic response belongs to another revision.");
  if (action === "diagnostic" && result.finding?.id !== options?.finding) throw new Error("Diagnostic response belongs to another finding.");
}
async function monitor(execute, poll, read, present, request, report, interval = 1000) {
  let finished = false, delivered = false, pending;
  const timer = setInterval(() => {
    if (finished || pending || delivered) return;
    pending = (async () => {
      try {
        const status = await poll();
        if (status.format !== STATUS || status.attempt_id !== request.attempt_id || status.request_id !== request.id) throw new Error("Delivery belongs to another request.");
        if (status.diagnostic_id && !finished) {
          const view = await read(status.diagnostic_id);
          if (view.request_id !== request.id || view.diagnostic_id !== status.diagnostic_id) throw new Error("Diagnostic revision belongs to another request.");
          if (!finished) { delivered = true; present(view); }
        }
      } catch (error) { if (!finished) report(`Early diagnostics unavailable: ${error.message}. Execution continues.`); }
      finally { pending = undefined; }
    })();
  }, interval);
  try { return await execute(); }
  finally { finished = true; clearInterval(timer); if (pending) await pending; }
}
module.exports = { VIEW, STATUS, GROUPS, argumentsFor, validate, monitor };
